lib/exclusive-lock.js

```diff
@@ -0,0 +1,157 @@
+      const err = new Error('Input validation failed')
+      err.code = 'EINVAL'
+      err.errors = validate.errors.map((e) => {
+        const {instancePath, message} = e
+        return {instancePath, message}
+      })
+      this.log.error(err)
+      throw err
+    }
+
+    const {
+      app_name
+    , log
+    , cache_connection
+    , lock_ttl_ms = DEFAULT_LOCK_TTL_MS
+    , lock_refresh_ms = DEFAULT_LOCK_REFRESH_MS
+    , lock_contents = 1 // Can be specified for more context, e.g. POD_NAME
+    , auto_refresh = true
+    } = opts
+
+    if (typeof cache_connection?.set !== 'function') {
+      const err = new TypeError('A valid cache_connection to a cache store is required')
+      err.code = 'EINVAL'
+      throw err
+    }
+    const diff = lock_ttl_ms - lock_refresh_ms
+    if (diff < REQUIRED_MIN_MS_DIFF) {
+      const err = new RangeError(
+        `lock_refresh_ms must be at least ${REQUIRED_MIN_MS_DIFF}ms less than lock_ttl_ms`
+      )
+      err.code = 'EINVAL'
+      err.meta = {
+        lock_ttl_ms
+      , lock_refresh_ms
+      , diff
+      }
+      throw err
+    }
+
+    this.cache_connection = cache_connection
+    this.app_name = app_name.replace(/[\s\W]+/g, '-')
+    this.log = log
+    this.lock_ttl_ms = lock_ttl_ms
+    this.lock_refresh_ms = lock_refresh_ms
+    this.lock_contents = lock_contents
+    this.auto_refresh = auto_refresh
+    this.#lock_name = `lock-manager:${this.app_name}`
+
+    this.got_lock = false
+    this.is_refreshing = false
+    this.refresh_timer = null
+  }
+
+  get lock_name() {
+    return this.#lock_name
+  }
+
+  async acquire() {
+    if (this.got_lock) {
+      this.log.warn('%s lock is already acquired. Must remove first.', this.lock_name)
+      return true
+    }
+    // We want this to throw if there are cache errors since that's a deal breaker
+    const got_lock = await this.cache_connection.set(
+      this.lock_name
+    , JSON.stringify(this.lock_contents)
+    , 'PX' // set expire in millis
+    , this.lock_ttl_ms
+    , 'NX' // set only if the record does Not eXist yet
+    )
+
+    if (got_lock) {
+      this.got_lock = true
+      if (this.auto_refresh) {
+        this.refresh_timer = setInterval(this.refresh.bind(this), this.lock_refresh_ms)
+      }
+      this.log.info({
+        lock_contents: this.lock_contents
+      , lock_refresh_ms: this.lock_refresh_ms
+      , lock_ttl_ms: this.lock_ttl_ms
+      }, '%s lock acquired', this.app_name)
+    }
+    return this.got_lock
+  }
+
+  async inspect() {
+    if (!this.got_lock) return
+    const raw = await this.cache_connection.get(this.lock_name)
+    try {
+      return JSON.parse(raw, jsonParseFilter)
+    } catch (err) {
+      this.log.error({err}, 'Lock contents is corrupt: %s', err)
+      return raw
+    }
+  }
+
+  refresh() {
+    if (!this.got_lock) return
+    if (this.is_refreshing) return
+    this.is_refreshing = true
+    this.cache_connection
+      .pexpire(this.lock_name, this.lock_refresh_ms)
```

Comment:
Yep, for sure.  What we have here is definitely just a first stab.  An enhancement like that would be good in the future. It would have to use the cache store to save state.

Reply:
I made an issue for this: https://github.com/logdna/exclusive-lock-node/issues/3

Reply:
Good call to capture it in a ticket 👍